receiver/postgresqlreceiver/client.go

```diff
@@ -66,13 +66,62 @@ type client interface {
 	getVersion(ctx context.Context) (string, error)
 	getQuerySamples(ctx context.Context, limit int64, logger *zap.Logger) ([]map[string]any, error)
 	getTopQuery(ctx context.Context, limit int64, logger *zap.Logger) ([]map[string]any, error)
+	explainQuery(query string, queryID string, logger *zap.Logger) (string, error)
 }
 
 type postgreSQLClient struct {
 	client  *sql.DB
 	closeFn func() error
 }
 
+// explainQuery implements client.
+func (c *postgreSQLClient) explainQuery(query string, queryID string, logger *zap.Logger) (string, error) {
+	_, err := c.client.Exec("/* otel-collector-ignore */ SET plan_cache_mode = force_generic_plan;")
+	if err != nil {
+		logger.Error("failed to set plan cache mode", zap.Error(err))
+		return "", err
+	}
+
+	normalizedqueryID := strings.ReplaceAll(queryID, "-", "_")
+	var queryBuilder strings.Builder
+	var nulls []string
+	counter := 1
+
+	for _, ch := range query {
+		if ch == '?' {
+			queryBuilder.WriteString(fmt.Sprintf("$%d", counter))
+			counter++
+			nulls = append(nulls, "null")
+		} else {
+			queryBuilder.WriteRune(ch)
+		}
+	}
+
+	preparedQuery := queryBuilder.String()
+
+	_, err = c.client.Exec(fmt.Sprintf("/* otel-collector-ignore */ PREPARE otel_%s AS %s", normalizedqueryID, preparedQuery))
+	if err != nil {
+		logger.Error("failed to create prepare statement", zap.Error(err), zap.String("query", preparedQuery))
+		return "", err
+	}
+	//nolint:errcheck
+	defer c.client.Exec(fmt.Sprintf("/* otel-collector-ignore */ DEALLOCATE PREPARE otel_%s", normalizedqueryID))
+
+	// if there is no parameter needed, we can not put an empty bracket
+	nullsString := ""
+	if len(nulls) > 0 {
+		nullsString = "(" + strings.Join(nulls, ", ") + ")"
+	}
+	wrappedDb := sqlquery.NewDbClient(sqlquery.DbWrapper{Db: c.client}, fmt.Sprintf("/* otel-collector-ignore */ EXPLAIN(FORMAT JSON) EXECUTE otel_%s%s", normalizedqueryID, nullsString), logger, sqlquery.TelemetryConfig{})
```

Comment:
It looks like we need to send many requests to get it done. is it possible to wrap them into one query?

Reply:
updated